### PR TITLE
Fix maxwell process DDL twice 

### PR DESCRIPTION
**Pull request by dongxiaohe**
Related to issue https://github.com/zendesk/maxwell/issues/780

Schema chain building conflicts with message processing on DDL.

In more details: 

What happen is that when Maxwell starts, it will compare the current position with schema binlog position and then build the schema linked list. This behaviour is correct and helps maxwell to reproduce all the db changes at any point of time. 

If the binlog position is DML, it is working fine. However, there is an edge case, if the binlog position is pointing to a DDL statement, maxwell would include this DDL statement into the schema chain. After that, maxwell will be unable to reprocess this statement again in `BinlogConnectorReplicator.processQueryEvent`. i.e. it couldn't create same table twice, it couldn't alter the same table twice and etc.

Solution: Use the parent schema from the latest if `binlog_position` and `target_position` are same.

Here are other solutions I can think of, but I feel like the drawback of it are pretty clear.
- Fix it in the `BinlogConnectorReplicator.processQueryEvent`: I don't think we should leave it at processing stage, it is very late and error prone.
- Point to previous binlog position: If it is DML, the same message would be reprocessing twice.

Any other ideas @zendesk/goanna  ?

@osheroff I am sorry to ask you review it, it is little abrupt, but this is my first code change to maxwell, I hope it won't break anything :)

### CODE REVIEW

**Comment by osheroff** [comment]
@dongxiaohe, was this something you encountered in production or just a theoretical repro?  I can kind of see happening in some worlds, but... hm... seems a bit unlikely.  For this to happen (and correct me if I'm wrong), it'd have to be:

- maxwell has already processed the DDL at `binlog:456`, but *not* advanced maxwell's position to `binlog:491`.  Now maxwell crashes, leaving the position pointer in place, and yeah, when we restart, we pick up the schema twice and things go wrong. 

To be clear: your fix appears correct, I'm just a bit dubious about the frequency of the edge case.

File: src/main/java/com/zendesk/maxwell/schema/MysqlSavedSchema.java [hunk 1]
```diff
@@ -573,7 +573,7 @@ private static Long findSchema(Connection connection, Position targetPosition, L
 			// Only consider binlog positions before the target position on the current server.
 			// Within those, sort for the latest binlog file, then the latest binlog position.
 			PreparedStatement s = connection.prepareStatement(
-				"SELECT id from `schemas` "
+				"SELECT id, binlog_position, base_schema_id from `schemas` "
 				+ "WHERE deleted = 0 "
 				+ "AND last_heartbeat_read <= ? AND ((binlog_file < ?) OR (binlog_file = ? and binlog_position <= ?)) AND server_id = ? "
```
Comment (osheroff) [inline src/main/java/com/zendesk/maxwell/schema/MysqlSavedSchema.java]: I think this whole change could actually be boiled down to changing `binlog_position <= ?` to `binlog_position < ?`, couldn't it?

**Comment by dongxiaohe** [comment]
@osheroff  Yeah, it happened in the production. After that, I can reproduce the issue locally via zdi.

File: src/main/java/com/zendesk/maxwell/schema/MysqlSavedSchema.java (hunk 1 above)
Comment (dongxiaohe) [reply src/main/java/com/zendesk/maxwell/schema/MysqlSavedSchema.java]: Nice !. Good thought, previously I thought this change may reproduce the same DML twice. On a second, this logic is only to find the latest schema change. I assume this should fix this edge case. I will amend the pull request and test it later on. 👍

File: src/test/java/com/zendesk/maxwell/MysqlSavedSchemaTest.java [hunk 2]
```diff
@@ -56,7 +57,9 @@ public void setUp() throws Exception {
 		this.position = MaxwellTestSupport.capture(server.getConnection());
 		this.context = buildContext(position);
 		this.schema = new SchemaCapturer(server.getConnection(), context.getCaseSensitivity()).capture();
-		this.savedSchema = new MysqlSavedSchema(this.context, this.schema, position);
+		this.savedSchema = new MysqlSavedSchema(this.context.getServerID(), this.context.getCaseSensitivity(),
```
Comment (dongxiaohe) [inline src/test/java/com/zendesk/maxwell/MysqlSavedSchemaTest.java]: This change is to fix all other tests.

File: src/test/java/com/zendesk/maxwell/SavedSchemaSupport.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+final public class SavedSchemaSupport {
+	private SavedSchemaSupport() { }
+
+	public static MysqlSavedSchema getSavedSchema(MaxwellContext context, Schema schema, Position position) throws Exception {
+		if (context.getConfig().gtidMode) {
+			return new MysqlSavedSchema(context, schema, position);
+		}
+
+		return new MysqlSavedSchema(context.getServerID(), context.getCaseSensitivity(),
+				schema, new Position(new BinlogPosition(position.getBinlogPosition().getOffset() - 1L,
```
Comment (osheroff) [inline src/test/java/com/zendesk/maxwell/SavedSchemaSupport.java]: sorry, I'm kinda confused;  why do we need the `getOffset() - 1` bits?

Comment (dongxiaohe) [reply src/test/java/com/zendesk/maxwell/SavedSchemaSupport.java]: Hi @osheroff , thanks for the question. This is only for `testing data`. The purpose of adding `- 1` is to make sure the seeding saved schema's position is prior to the current position. It fixed lots of failed tests in `MysqlSavedSchemaTest` and `RecoveryTest` (Previously, the saved schema is same position with current position. (lots of null pointer exception in the tests)). :)

Comment (osheroff) [reply src/test/java/com/zendesk/maxwell/SavedSchemaSupport.java]: ok.  it's a little nasty, but ok.
